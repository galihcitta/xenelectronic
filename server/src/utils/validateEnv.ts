import { port, cleanEnv } from 'envalid';

export const validateEnv = () => {
  cleanEnv(process.env, {
    PORT: port()
  });
};

// https://www.npmjs.com/package/envalid
