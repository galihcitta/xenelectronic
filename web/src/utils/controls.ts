import { Controls } from '../pages/Auth/types';

export const signInControls: Controls = {
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signin-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signin-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};

export const signUpControls: Controls = {
  username: {
    label: 'Enter username',
    placeholder: '',
    type: 'text',
    id: 'signup-username',
    name: 'username',
    value: '',
    validationRules: {
      required: true,
      minLength: 4,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signup-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signup-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true,
      isSecure: true,
      minLength: 8,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};
