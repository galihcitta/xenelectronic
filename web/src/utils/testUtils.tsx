import React from 'react';
import { BrowserRouter, Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../modules/store';
import { createMemoryHistory } from 'history';

const history = createMemoryHistory();

export const withRouter = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <BrowserRouter>
    <Component {...props} />
  </BrowserRouter>
);

export const withHistory = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Router history={history}>
    <Component {...props} />
  </Router>
);

export const withProvider = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Provider store={store}>
    <BrowserRouter>
      <Component {...props} />
    </BrowserRouter>
  </Provider>
);
