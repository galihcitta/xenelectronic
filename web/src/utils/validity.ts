import { ValidationRules } from '../pages/Auth/types';

export const validate = (
  value: string,
  controlName: string,
  rules: ValidationRules
) => {
  let isValid = true;
  let errorMsg = '';

  if (rules.required) {
    isValid = value.trim() !== '' && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} field is required`;
  }

  if (rules.minLength) {
    isValid = value.length >= rules.minLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be at least ${rules.minLength} characters`;
  }

  if (rules.maxLength) {
    isValid = value.length <= rules.maxLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be no longer than ${rules.maxLength} characters`;
  }

  if (rules.isEmail) {
    const pattern = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `email is invalid`;
  }

  if (rules.isSecure) {
    // Minimum eight characters, at least one uppercase letter, one lowercase letter, one number and one special character
    const pattern = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `password too weak`;
  }

  return {
    isValid,
    errorMsg: errorMsg.charAt(0).toUpperCase() + errorMsg.slice(1)
  };
};
