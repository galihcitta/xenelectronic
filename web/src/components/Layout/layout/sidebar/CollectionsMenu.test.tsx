import { render, fireEvent } from '@testing-library/react';
import {
  _CollectionsMenu,
  CollectionsMenuProps
} from './CollectionsMenu.compnent';
import { withRouter } from '../../../../utils/testUtils';

const CollectionMenuWithRouter = withRouter(_CollectionsMenu);

describe('<CollectionsMenu />', () => {
  const mockCloseInterior = jest.fn();
  const mockCloseExterior = jest.fn();
  const mockHandleCloseMenu = jest.fn();
  const mockFetchProductsByType = jest.fn();

  const props: CollectionsMenuProps = {
    showInterior: false,
    showExterior: false,
    closeInterior: mockCloseInterior,
    closeExterior: mockCloseExterior,
    handleCloseMenu: mockHandleCloseMenu,
    fetchProductsByType: mockFetchProductsByType
  };

  const setup = (props: CollectionsMenuProps) => {
    const utils = render(CollectionMenuWithRouter(props));
    return { ...utils };
  };

  it('renders correctly', () => {
    const { asFragment } = setup(props);
    expect(asFragment()).toMatchSnapshot();
  });

  it('calls `handleCloseMenu` func', () => {
    const { getByText } = setup(props);
    fireEvent.click(getByText('Clocks'));

    expect(mockHandleCloseMenu).toHaveBeenCalledTimes(1);
  });

  it('calls `fetchProductsByType` action with `clocks` arg', () => {
    const { getByText } = setup(props);
    fireEvent.click(getByText('Clocks'));

    expect(mockFetchProductsByType).toBeCalledWith('clocks');
  });
});
