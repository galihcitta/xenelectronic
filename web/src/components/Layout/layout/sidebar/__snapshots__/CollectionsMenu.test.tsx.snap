// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CollectionsMenu /> renders correctly 1`] = `
<DocumentFragment>
  <nav
    class="sc-EHOje jWYlbq"
    data-testid="Wrapper"
  >
    <header
      class="sc-bZQynM lpjVvd"
    >
      <p
        class="sc-gzVnrw dNIQQe"
      >
        Interior
      </p>
      <button
        aria-label="close"
        class="sc-htoDjs qBCwI"
        title="close menu"
      >
        <i
          class="far fa-times-circle"
        />
      </button>
    </header>
    <ul
      class="sc-dnqmqq eBryZD"
    >
      <li
        class="sc-bdVaJa jqTRNv"
      >
        <a
          class="sc-ifAKCX fYSqpX"
          href="/shop"
        >
          Chairs
        </a>
      </li>
      <li
        class="sc-bdVaJa jqTRNv"
      >
        <a
          class="sc-ifAKCX fYSqpX"
          href="/shop"
        >
          Lamps
        </a>
      </li>
      <li
        class="sc-bdVaJa jqTRNv"
      >
        <a
          class="sc-ifAKCX fYSqpX"
          href="/shop"
        >
          Clocks
        </a>
      </li>
      <li
        class="sc-bdVaJa jqTRNv"
      >
        <a
          class="sc-ifAKCX fYSqpX"
          href="/shop"
        >
          Vases
        </a>
      </li>
    </ul>
  </nav>
  <nav
    class="sc-EHOje jWYlbq"
    data-testid="Wrapper"
  >
    <header
      class="sc-bZQynM lpjVvd"
    >
      <p
        class="sc-gzVnrw dNIQQe"
      >
        Exterior
      </p>
      <button
        aria-label="close"
        class="sc-htoDjs qBCwI"
        title="close menu"
      >
        <i
          class="far fa-times-circle"
        />
      </button>
    </header>
    <ul
      class="sc-dnqmqq eBryZD"
    >
      <li
        class="sc-bdVaJa jqTRNv"
      >
        <a
          class="sc-ifAKCX fYSqpX"
          href="/shop"
        >
          Benches
        </a>
      </li>
      <li
        class="sc-bdVaJa jqTRNv"
      >
        <a
          class="sc-ifAKCX fYSqpX"
          href="/shop"
        >
          Flowerpots
        </a>
      </li>
      <li
        class="sc-bdVaJa jqTRNv"
      >
        <a
          class="sc-ifAKCX fYSqpX"
          href="/shop"
        >
          Plants
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
